crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2579 @@
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P384)) {
+                kem = OSSL_HPKE_KEM_ID_P384;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P521)) {
+                kem = OSSL_HPKE_KEM_ID_P521;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_X25519)) {
+                kem = OSSL_HPKE_KEM_ID_X25519;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_X448)) {
+                kem = OSSL_HPKE_KEM_ID_X448;
+            }
+            if (HPKE_MSMATCH(st, "0x10")) { kem = OSSL_HPKE_KEM_ID_P256; }
+            if (HPKE_MSMATCH(st, "16")) { kem = OSSL_HPKE_KEM_ID_P256; }
+            if (HPKE_MSMATCH(st, "0x11")) { kem = OSSL_HPKE_KEM_ID_P384; }
+            if (HPKE_MSMATCH(st, "17")) { kem = OSSL_HPKE_KEM_ID_P384; }
+            if (HPKE_MSMATCH(st, "0x12")) { kem = OSSL_HPKE_KEM_ID_P521; }
+            if (HPKE_MSMATCH(st, "18")) { kem = OSSL_HPKE_KEM_ID_P521; }
+            if (HPKE_MSMATCH(st, "0x20")) { kem = OSSL_HPKE_KEM_ID_X25519; }
+            if (HPKE_MSMATCH(st, "32")) { kem = OSSL_HPKE_KEM_ID_X25519; }
+            if (HPKE_MSMATCH(st, "0x21")) { kem = OSSL_HPKE_KEM_ID_X448; }
+            if (HPKE_MSMATCH(st, "33")) { kem = OSSL_HPKE_KEM_ID_X448; }
+        } else if (kem != 0 && kdf == 0) {
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KDFSTR_256)) { kdf = 1; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KDFSTR_384)) { kdf = 2; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KDFSTR_512)) { kdf = 3; }
+            if (HPKE_MSMATCH(st, "0x01")) { kdf = 1; }
+            if (HPKE_MSMATCH(st, "0x02")) { kdf = 2; }
+            if (HPKE_MSMATCH(st, "0x03")) { kdf = 3; }
+            if (HPKE_MSMATCH(st, "0x1")) { kdf = 1; }
+            if (HPKE_MSMATCH(st, "0x2")) { kdf = 2; }
+            if (HPKE_MSMATCH(st, "0x3")) { kdf = 3; }
+            if (HPKE_MSMATCH(st, "1")) { kdf = 1; }
+            if (HPKE_MSMATCH(st, "2")) { kdf = 2; }
+            if (HPKE_MSMATCH(st, "3")) { kdf = 3; }
+        } else if (kem != 0 && kdf != 0 && aead == 0) {
+            if (HPKE_MSMATCH(st, OSSL_HPKE_AEADSTR_AES128GCM)) { aead = 1; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_AEADSTR_AES256GCM)) { aead = 2; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_AEADSTR_CP)) { aead = 3; }
+            if (HPKE_MSMATCH(st, "0x01")) { aead = 1; }
+            if (HPKE_MSMATCH(st, "0x02")) { aead = 2; }
+            if (HPKE_MSMATCH(st, "0x03")) { aead = 3; }
+            if (HPKE_MSMATCH(st, "0x1")) { aead = 1; }
+            if (HPKE_MSMATCH(st, "0x2")) { aead = 2; }
+            if (HPKE_MSMATCH(st, "0x3")) { aead = 3; }
+            if (HPKE_MSMATCH(st, "1")) { aead = 1; }
+            if (HPKE_MSMATCH(st, "2")) { aead = 2; }
+            if (HPKE_MSMATCH(st, "3")) { aead = 3; }
+        }
+        st = strtok(NULL, ",");
+    }
+    OPENSSL_free(instrcp);
+    if ((st != NULL && labels > 3) || kem == 0 || kdf == 0 || aead == 0) {
+        return 0;
+    }
+    suite->kem_id = kem;
```

Comment:
Should this also fail if its not a valid combination

Reply:
All current combinations are valid, both in protocol terms and for this implementation. Even if that changed in future, I think it'd make more sense to not fail the string mapping but later when some non-supported suite is attempted, so I'd say better to not add such a check here.